kedro-org/kedro
# Allow users to pass credentials through environment variables

> **NOTE:** Kedro datasets are moving from `kedro.extras.datasets` to a separate `kedro-datasets` package in
> [`kedro-plugins` repository](https://github.com/kedro-org/kedro-plugins). Any changes to the dataset implementations
> should be done by opening a pull request in that repository.
## Description

Resolves [#1909](https://github.com/kedro-org/kedro/issues/1909).

## Development notes

`OmegaConf` resolves `DictConfig` objects on read. In the following example (where credentials.yml contains environment variable grammar such as `"key": "${oc.env:TEST_KEY}"`), the `oc.env` resolver would have to be registered, or line 3 will raise an `UnsupportedInterpolationType` exception.

```
conf = OmegaConfLoader(...)
credentials = conf["credentials"]
credentials["test"]
```

However, the scope of the issue is to keep the `oc.env` resolver switched off. The way around this is to use `OmegaConf.resolve`, which resolves `DictConfig` objects in-place when these objects are created, using the `oc.env` resolver. This is done in the `_resolve_environment_variables` method.

Includes four automated tests:

1. Check that environment variables are resolved when the `credentials` is used as a key.
2. Check that environment variables are not resolved when `credentials` is not used as a key.
3. Check that the `oc.env` resolver is not registered after environment variables are read if it was not registered beforehand.
4. Check that the `oc.env` resolver remains registered after environment variables are read if it was registered beforehand.

## Checklist



<a href="https://gitpod.io/#https://github.com/kedro-org/kedro/pull/2178"><img src="https://gitpod.io/button/open-in-gitpod.svg"/></a>
